contracts/MarginCalculator.sol

```diff
@@ -43,19 +36,227 @@ contract MarginCalculator {
+     * @param _strike otoken strike asset
+     * @param _collateral otoken collateral asset
+     * @param _isPut otoken type
+     * @param _shockValue spot shock value
+     */
+    function setSpotShock(
+        address _underlying,
+        address _strike,
+        address _collateral,
+        bool _isPut,
+        uint256 _shockValue
+    ) external onlyOwner {
+        bytes32 productHash = keccak256(abi.encode(_underlying, _strike, _collateral, _isPut));
+
+        spotShock[productHash] = _shockValue;
+    }
+
+    /**
+     * @notice set oracle deviation (1e27)
+     * @dev can only be called by owner
+     * @param _deviation deviation value
+     */
+    function setOracleDeviation(uint256 _deviation) external onlyOwner {
+        oracleDeviation = _deviation;
+    }
+
+    /**
+     * @notice get dust amount for collateral asset
+     * @param _collateral collateral asset address
+     * @return dust amount (1e27)
+     */
+    function getCollateralDust(address _collateral) external view returns (uint256) {
+        return dust[_collateral];
+    }
+
+    /**
+     * @notice get time to expiry for specific product
+     * @param _underlying otoken underlying asset
+     * @param _strike otoken strike asset
+     * @param _collateral otoken collateral asset
+     * @param _isPut otoken type
+     * @return array of times to expiry
+     */
+    function getProductTimeToExpiry(
+        address _underlying,
+        address _strike,
+        address _collateral,
+        bool _isPut
+    ) external view returns (uint256[] memory) {
+        bytes32 productHash = keccak256(abi.encode(_underlying, _strike, _collateral, _isPut));
+        return productTimeToExpiry[productHash];
+    }
+
+    /**
+     * @notice get option upper bound value for specific time to expiry
+     * @param _underlying otoken underlying asset
+     * @param _strike otoken strike asset
+     * @param _collateral otoken collateral asset
+     * @param _isPut otoken type
+     * @param _timeToExpiry option time to expiry timestamp
+     * @return option upper bound value (1e27)
+     */
+    function getTimeToExpiryValue(
+        address _underlying,
+        address _strike,
+        address _collateral,
+        bool _isPut,
+        uint256 _timeToExpiry
+    ) external view returns (uint256) {
+        bytes32 productHash = keccak256(abi.encode(_underlying, _strike, _collateral, _isPut));
+
+        return timeToExpiryValue[productHash][_timeToExpiry];
+    }
+
+    /**
+     * @notice get spot shock value
+     * @param _underlying otoken underlying asset
+     * @param _strike otoken strike asset
+     * @param _collateral otoken collateral asset
+     * @param _isPut otoken type
+     * @return _shockValue spot shock value (1e27)
+     */
+    function getSpotShock(
+        address _underlying,
+        address _strike,
+        address _collateral,
+        bool _isPut
+    ) external view returns (uint256) {
```

Comment:
Does it need to external?

Reply:
Yeah, it is a getter function know what is the spot shock value for a product is, as `spotShock` mapping is internal.

Reply:
Added.